ex6p built with XL`-O3 -qhot -qipa` produces incorrect results

On the [RPI DCS system (2 P9 + 4 GV100 per node)](https://secure.cci.rpi.edu/wiki/index.php?title=DCS_Supercomputer) I built mfem and its dependencies as detailed below.

I first built mfem and `ex6p` with the default XL compiler flags `-O3 -qhot -qipa`.  Running `./ex6p -pa -d cuda` on four processes, one gpu per process, completes one AMR cycle and exits.  `stderr` and `stdout` is pasted below:

```
Options used:
   --mesh ../data/star.mesh
   --order 1
   --partial-assembly
   --device cuda
   --visualization
Device configuration: cuda,cpu
Unable to connect to GLVis server at localhost:19916
GLVis visualization disabled.

AMR iteration 0
Number of unknowns: 31
   Iteration :   0  (B r, r) = 0.653648 ... 
   Iteration :   3  (B r, r) = 9.11866e-14
Average reduction factor = 0.00720164
Stopping criterion satisfied. Stop.
```

Buidling with `-g -O2` (which override the default optimization level) and running `./ex6p -pa -d cuda` on four processes, one gpu per process, produces the following correct output:

```
Options used:
   --mesh ../data/star.mesh
   --order 1
   --partial-assembly
   --device cuda
   --visualization
Device configuration: cuda,cpu
Unable to connect to GLVis server at localhost:19916
GLVis visualization disabled.

AMR iteration 0
Number of unknowns: 31
   Iteration :   0  (B r, r) = 0.653648 ...
   Iteration :   3  (B r, r) = 9.11866e-14
Average reduction factor = 0.00720164

AMR iteration 1
Number of unknowns: 101 
   Iteration :   0  (B r, r) = 0.124108 ...
   Iteration :   9  (B r, r) = 7.83064e-15
Average reduction factor = 0.184785

<.. snip ..>

AMR iteration 26
Number of unknowns: 123036
   Iteration :   0  (B r, r) = 7.30948e-06 ...
   Iteration : 314  (B r, r) = 7.2351e-18
Average reduction factor = 0.95694
Elapsed time <rank> <ms> 0 0.24462
Reached the maximum number of dofs. Stop.
```

## Setup

`module load xl_r cuda mpich`

## Hypre

```
wget  https://github.com/hypre-space/hypre/archive/v2.16.0.tar.gz
ln -s hypre-2.16.0 hypre
cd hypre/src
./configure --disable-fortran --with-MPI --with-MPI-include=/gpfs/u/software/ppc64le-rhel7/.xl_r/16/mpich/3.2.1/1//include --with-MPI-lib-dirs=/gpfs/u/software/ppc64le-rhel7/.xl_r/16/mpich/3.2.1/1//lib
make -j 8
```

## Metis

```
wget http://glaros.dtc.umn.edu/gkhome/fetch/sw/metis/metis-5.1.0.tar.gz
sed -i 's/O3/02/g' GKlib/GKlibSystem.cmake  # I suspect there are problems with default compiler optimization levels, see below
CC=xlc_r make config prefix=`pwd`/../build-metis510-xl_r
make -j 8 install
```

## MFEM

```
module load xl_r mpich cuda cmake
git clone git@github.com:mfem/mfem.git
cd mfem
sed -i 's/OPTIM_FLAGS = -O3/OPTIM_FLAGS = -O2/g' config/defaults.mk
sed -i 's/DEBUG_FLAGS =/DEBUG_FLAGS = -O2/g' config/defaults.mk
make -j 8 pcudebug CUDA_ARCH=sm_70 MFEM_USE_METIS_5=YES METIS_DIR=/gpfs/u/barn/ITAP/ITAPsmth/mfem/../build-metis510-xl_r
cd examples
make ex6p
```

Using the `pcudebug` configuration adds `-g` which does not override the XL compilers default `-O3 -qhot -qipa` flags.

## Run

https://secure.cci.rpi.edu/wiki/index.php?title=DCS_Supercomputer#Example:_running_with_one_process_per_GPU

run.sh

```
#!/bin/sh
#SBATCH -t 10
#SBATCH --partition dcs
#SBATCH --nodes 2
#SBATCH --gres=gpu:4

module load xl_r mpich cuda
echo $CUDA_VISIBLE_DEVICES
set -x
srun --gres=gpu:4 -n 8 ./bindProcessToGpu.sh ./ex6p -pa -d cuda
set +x
```

bindProcessToGpu.sh

```
#!/bin/bash
export CUDA_VISIBLE_DEVICES=$SLURM_LOCALID
$@
```

submit

`sbatch ./run.sh`

I set `OPTIM_FLAGS = -qstrict -O3 $(BASE_FLAGS)` in `config/default.mk` then ran:
`make -j 8 pcuda CUDA_ARCH=sm_70 MFEM_USE_METIS_5=YES METIS_DIR=/gpfs/u/barn/ITAP/ITAPsmth/mfem/../build-metis510-xl_r`.  

`nvcc` reports `nvcc fatal   : Unknown option 'qstrict'` .  Is there a way around this?

The full output is below.

```
make -f makefile config MFEM_USE_MPI=YES MFEM_DEBUG=NO \
   MFEM_USE_CUDA=YES MFEM_USE_MM=YES METIS_DIR=/gpfs/u/barn/ITAP/ITAPsmth/mfem/../build-metis510-xl_r MFEM_USE_METIS_5=YES CUDA_ARCH=sm_70
make[1]: Entering directory `/gpfs/u/barn/ITAP/ITAPsmth/mfem'
make -C config all
make[2]: Entering directory `/gpfs/u/barn/ITAP/ITAPsmth/mfem/config'
nvcc -I../../hypre/src/hypre/include get_hypre_version.cpp -o get_hypre_version
./get_hypre_version > get_hypre_version.out
HYPRE version: 21600
rm get_hypre_version get_hypre_version.out
make[2]: Leaving directory `/gpfs/u/barn/ITAP/ITAPsmth/mfem/config'

Build destination: <source> [/gpfs/u/barn/ITAP/ITAPsmth/mfem]

make[1]: Leaving directory `/gpfs/u/barn/ITAP/ITAPsmth/mfem'
make METIS_DIR=/gpfs/u/barn/ITAP/ITAPsmth/mfem/../build-metis510-xl_r MFEM_USE_METIS_5=YES CUDA_ARCH=sm_70
make[1]: Entering directory `/gpfs/u/barn/ITAP/ITAPsmth/mfem'
nvcc   -qstrict -O3 -std=c++11 -x=cu --expt-extended-lambda -arch=sm_70 -ccbin mpicxx -I./../hypre/src/hypre/include -I/gpfs/u/barn/ITAP/ITAPsmth/mfem/../build-metis510-xl_r/include  -c general/optparser.cpp -o general/optparser.o
nvcc fatal   : Unknown option 'qstrict'
make[1]: *** [general/optparser.o] Error 1
make[1]: Leaving directory `/gpfs/u/barn/ITAP/ITAPsmth/mfem'
make: *** [pcuda] Error 2
```

--- a/mesh/mesh_operators.cpp
+++ b/mesh/mesh_operators.cpp
@@ -98,9 +98,16 @@ int ThresholdRefiner::ApplyImpl(Mesh &mesh)
    const double total_err = GetNorm(local_err, mesh);
    if (total_err <= total_err_goal) { return STOP; }
 
-   const double exp = (total_norm_p < infinity()) ? -1.0/total_norm_p : 0.0;
-   const double ne_inorm = std::pow(num_elements, exp);
-   threshold = std::max(total_err * total_fraction * ne_inorm, local_err_goal);
+   if (total_norm_p < infinity())
+   {
+      threshold = std::max(total_err * total_fraction *
+                           std::pow(num_elements, -1.0/total_norm_p),
+                           local_err_goal);
+   }
+   else
+   {
+      threshold = std::max(total_err * total_fraction, local_err_goal);
+   }
 
    for (int el = 0; el < NE; el++)
    {

--- a/mesh/mesh_operators.cpp
+++ b/mesh/mesh_operators.cpp
@@ -95,12 +95,12 @@ int ThresholdRefiner::ApplyImpl(Mesh &mesh)
    const Vector &local_err = estimator.GetLocalErrors();
    MFEM_ASSERT(local_err.Size() == NE, "invalid size of local_err");
 
-   double total_err = GetNorm(local_err, mesh);
+   const double total_err = GetNorm(local_err, mesh);
    if (total_err <= total_err_goal) { return STOP; }
 
-   threshold = std::max(total_err * total_fraction *
-                        std::pow(num_elements, -1.0/total_norm_p),
-                        local_err_goal);
+   const double exp = (total_norm_p < infinity()) ? -1.0/total_norm_p : 0.0;
+   const double ne_inorm = std::pow(num_elements, exp);
+   threshold = std::max(total_err * total_fraction * ne_inorm, local_err_goal);
 
    for (int el = 0; el < NE; el++)
    {

